Fix some vararg-related subtyping issues

Keno opened the pull request:
This is a backport from (a local version of) #31681. Turns out the same subtyping issues I addressed there exist on master, but we just tend to not see those types a whole lot.

Keno committed:
Fix jl_obvious_subtype with INT vararg constraint

Keno committed:
Fix a vararg-related non-transitivity in subtyping

JeffBezanson commented:
Rebased on master to pick up the gc root fix.

Keno commented:
I have more cases to fix that partially correct what's here, so let's do them all at once.

Keno committed:
Fix another vararg subtype issue

Keno commented:
Current set of tests, I'm trying to fix, beyond what's in this PR:
```
let A = Tuple{Vararg{Val{N}, N} where N},
    B = Tuple{Vararg{Val{N}, N}} where N,
    C = Tuple{Val{2}, Val{2}}

    @test A == B
    @test issub(B, A)
    @test issub(C, A)
end
@test Tuple{T, Vararg{T, 2}} == Tuple{Vararg{T, 3}}
```

Keno committed:
Take advantage of their being at most one UnionAll wrapped around a Vararg

Upon construction, we normalize `Vararg{T, N} where {T,N}` to `Vararg{T where T, N} where N`,
thus there can be at most one UnionAll wrapper around a Vararg. In subtyping we were already
assuming that there can be at most two such wrappers, so simply adjust that and add an
appropriate assertion to make sure we catch any cases where this ever goes wrong.

Keno committed:
Rewrite subtype_tuple to fix extra cases

Keno commented:
Note to self: The following is still wrong on this branch:
```
(Tuple{Vararg{T, 10}} where T<:Real) <: Base.Any16
```

Keno commented:
Looks like this branch also fixes the following from #19998:
```
julia> (Ref{Tuple{Int, Vararg{Int, N}}} where N) <: (Ref{Tuple{Vararg{Int, N}}} where N)
true
```

Keno commented:
Looks like I need to put the vararg subtyping check back for things like `Type{Vararg} == Type{Vararg}` which does get asked about when Vararg is used explicitly in the code.

JeffBezanson commented:
We should try to avoid that, since forming those types should not be allowed.

Keno commented:
It comes up in situations like the following:
```
push!(Any[], Vararg)
```
for which the compiled signature is `Tuple{typeof(push!), Vector{Any}, Type{Vararg}}`. I guess we could disallow `Vararg` from `Typeof`.

Keno committed:
Put back the case for naked varargs

Keno commented:
@JeffBezanson This is ready for re-review.

Keno commented:
Linking #30995 here for the `Type{Vararg}` issue. I guess the conclusion is that that isn't legal, so we should find out who forms it and fix it.

vtjnash commented on src/subtype.c:
```diff
@@ -1301,24 +1483,37 @@ JL_DLLEXPORT int jl_obvious_subtype(jl_value_t *x, jl_value_t *y, int *subtype)
-            if (npx != npy || vx || vy) {
-                if (!vy) {
+            if (npx != npy || vx != JL_VARARG_NONE || vy) {
+                if ((vx == JL_VARARG_NONE || vx == JL_VARARG_UNBOUND) && !vy) {
```
I think `UNBOUND` is computed incorrectly for subtyping's use (it was introduced for gf.c's heuristic type signature limiting), since it doesn't count number of uses. Consider `Tuple{Vararg{NTuple{N}, N} where N}`. To be conservative, just consider `jl_is_long` vs `jl_is_typevar` in this fast-path.

Keno replied:
What the problem with that case? If the RHS is not a vararg, then if we have a unionall here it can't possibly be a subtype.

vtjnash replied:
I don't know. I just don't trust this code to do any counting of TypeVar uses.

vtjnash commented on test/subtype.jl:
```diff
@@ -1171,7 +1173,7 @@ end
 f20103(::Type{TT20103{X,Y}},x::X,y::Y) where {X,Y} = 1
 f20103(::Type{TT20103{X,X}},x::X) where {X} = 100
-@test_broken typeintersect(Type{NTuple{N,E}} where E where N, Type{NTuple{N,E} where N} where E) == Union{} # use @testintersect once fixed
+@test typeintersect(Type{NTuple{N,E}} where E where N, Type{NTuple{N,E} where N} where E) == Union{} # use @testintersect once fixed
```
```suggestion
@testintersect(Type{NTuple{N,E}} where E where N, Type{NTuple{N,E} where N} where E, Union{})
```

vtjnash commented on src/subtype.c:
```diff
@@ -796,90 +818,186 @@ static int check_vararg_length(jl_value_t *v, ssize_t n, jl_stenv_t *e)
+static int subtype_tuple_tail(struct subtype_tuple_env *env, int8_t R, jl_stenv_t *e, int param)
+{
+loop:
+    // while (i <= lx) {
```
maybe put this comment on the previous line, so be clearer that it's a comment and not deadcode?

vtjnash commented on src/subtype.c:
```diff
@@ -796,90 +818,186 @@ static int check_vararg_length(jl_value_t *v, ssize_t n, jl_stenv_t *e)
-                    if (ylv && jl_is_long(ylv->lb) && jl_unbox_long(ylv->lb) == 0)
-                        break;
+            xi = env->vtx;
+        } else {
```
```suggestion
        } 
        else {
```
(everywhere)

JeffBezanson commented:
I think this fixes #31805? Would be good to add that test case.

vtjnash committed:
Update test/subtype.jl

Co-Authored-By: Keno <keno@alumni.harvard.edu>

Keno committed:
Add test for #31805

Keno committed:
Fix style review comments

Keno committed:
Rename variable

JeffBezanson commented on base/compiler/typelimits.jl:
```diff
@@ -92,10 +92,13 @@ function _limit_type_size(@nospecialize(t), @nospecialize(c), sources::SimpleVec
-        return t # t isn't something new
+    else
+        ut = unwrap_unionall(t)
+        if isa(ut, DataType) && ut.name !== unwrap_unionall(Vararg).name && isa(c, Type) && c !== Union{} && c <: t
```
Can use `Base._va_typename` here.

Keno replied:
(Core.Compiler.) `_va_typename`, but yes

Keno committed:
In person review changes

Keno committed:
Fix bug

Keno commented:
PkgEval reveals the following incorrect intersection on this branch:
```
typeintersect(Tuple{Array{Tuple{Vararg{Int64,N}},N},Tuple{Vararg{Array{Int64,1},N}}} where N, Tuple{Array{Tuple{Int64},1}, Tuple})
```
(`Tuple{Array{Tuple{Int64},1},Tuple{Array{Int64,1}}}`) is in the intersection.

Keno committed:
Handle integer bounds on left arguments in the environment

In subtyping proper, variables introduced on the left (i.e. forall
variables) don't have any equality constraints, because we have no
syntax for creating them. However, intersection does sometimes create
such environments, so we need to handle it in subtyping.

Keno commented:
PkgEval is clean now. Will merge.

Keno commented:
Let's backport this to fix #32607

ararslan commented:
Running PkgEval with the backports branch suggests that this didn't fix the issue
